lib/util.ts

```diff
@@ -112,31 +125,63 @@ export class BroadcastUtil {
    * @returns The result tensor, or undefined if input not broadcastable.
    */
   static calc(a: ndarray, b: ndarray, op: (a: number, b: number) => number): ndarray|undefined {
-    const shape = BroadcastUtil.calcShape(a.shape, b.shape);
-    if (shape) {
-      const size = ShapeUtil.size(shape);
+    const outputShape = BroadcastUtil.calcShape(a.shape, b.shape);
+
+    if (outputShape) {
+      const size = ShapeUtil.size(outputShape);
       const c = ndarray(
           new (
               a.data.constructor as Int8ArrayConstructor | Int16ArrayConstructor | Int32ArrayConstructor |
               Uint8ArrayConstructor | Uint16ArrayConstructor | Uint32ArrayConstructor | Float32ArrayConstructor |
               Float64ArrayConstructor | Uint8ClampedArrayConstructor)(size),
-          shape as number[]);
-
-      const indices = new Array<number>(shape.length);
-      for (let i = 0; i < size; i++) {
-        // traversal indices
-        let rest = i;
-        for (let j = shape.length - 1; j >= 0; j--) {
-          indices[j] = rest % shape[j];
-          rest = Math.floor(rest / shape[j]);
+          outputShape as number[]);
+
+      // both inputs are scalars
+      if (outputShape.length === 0) {
+        c.set(...[], op(a.get(...[]), b.get(...[])));
+      }
+
+      // atleast one input is a non-scalar
+      else {
+        const outputIndices = new Array<number>(outputShape.length);
+        const originalIndicesA = new Array(a.shape.length);
+        const originalIndicesB = new Array(b.shape.length);
+        let valA = 0;
+        let valB = 0;
+        let isAScalar = false;
+        let isBScalar = false;
+        if (a.shape.length === 0) {
+          valA = a.get(...[]);
+          isAScalar = true;
+        }
+        if (b.shape.length === 0) {
+          valB = b.get(...[]);
+          isBScalar = true;
         }
+        let rest: number;
```

Comment:
why do we need this `rest` instead of using `i`

Reply:
We need `rest` as that value gets modified in the inner loop. This is no different from existing implementation, only that rest has been declared just once outside the outer loop...